lib/bdev/nvme/blockdev_nvme.c

```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
 }
 
 static void
-blockdev_nvme_timeout_cb(void *cb_arg, struct spdk_nvme_ctrlr *ctrlr,
-			 struct spdk_nvme_qpair *qpair, uint16_t cid)
+abort_cb(void *arg, const struct spdk_nvme_cpl *cpl)
+{
+	/*
+	 * abort completion callback function. Currently, there is nothing to do
+	 * struct spdk_nvme_ctrlr *ctrlr = arg;
+	 */
+}
+
+static void
+blockdev_nvme_timeout_cb(struct spdk_nvme_ctrlr *ctrlr,
+			 struct spdk_nvme_qpair *qpair,
+			 void *cb_arg,
```

Comment:
As I commented previously, when a command times out, the timeout value is extended by refreshing tr->submit_ticks before calling the callback. This extension has two purposes. It prevents calling the callback over and over again on subsequent polls and it provides a fallback in case of abort failure.

At the callback function, timeout_count will either be 1 indicating that this is the initial timeout, or 2 indicating that the extended time has also elapsed.

At the callback with aborts enable:
If timeout_count = 1, the callback will issue an abort. 
If timeout_count > 1, this indicates that an abort was sent and the original I/O still did not complete within the extended time period. Therefore, the callback will reset the controller.

We actually do not do any error checking for the abort command other than checking for timeout (which results in an automatic reset.) All other errors will be covered by the timeout extension. (We actually had earlier code that avoided the "abort command limit exceeded" scenario - we can add this at a later PR.)